lib/TTT/player.ex

```diff
@@ -0,0 +1,58 @@
+defmodule Player do
+  def available_moves(state) do
+    board = State.get_board(state)
+    for {key, value} <- board, is_number(value), into: %{}, do: {key, value}
+  end
+
+  def select_move(state) do
+    try do
+      available_moves(state)
+      |> Map.keys()
+      |> Enum.random()
+    rescue
+      e in Enum.EmptyError -> {:error, e.message}
+    end
+  end
+
+  def choose_token(deps) do
+    token =
+      deps.io.retrieve(deps.messages.get_message(:choose_token))
+      |> deps.validation.clean()
+      |> String.upcase()
+
+    case token do
+      "X" ->
+        token
+
+      "O" ->
+        token
+
+      _ ->
+        choose_token(deps)
+    end
+  end
+
+  def analyze(move, state, deps) do
+    with {:ok, :space_on_board} <- deps.validation.space_on_board?(move),
+         {:ok, :is_open} <- deps.validation.open?(move, state) do
+      place_move(move, state)
+    else
+      {:error, :nonexistant_space} ->
```

Comment:
Oh hey, I didn't know you could pattern match on the `else` of a `with`.  Cool!

Reply:
Yeah! After working in Elixir for a little bit I want `with` along with the pipe operator in every language